spec/active_yaml/aliases_spec.rb

```diff
@@ -56,10 +61,15 @@ class KeyProduct < ActiveYaml::Base
     end
 
     describe 'keys starting with "/"' do
-      it 'excludes them' do
+      it 'excludes them from records' do
         models_including_aliases = model.all.select { |p| p.attributes.keys.include? :'/aliases' }
         expect(models_including_aliases).to be_empty
       end
+
+      it 'excludes them from fields' do
+        model.all
+        expect(model.field_names).to match_array [:name, :flavor, :price, :slogan]
+      end
```

Comment:
It worked in hash style, thus this test passes with or without the change.

Reply:
I'm a little confused how this passes both before and after when you sample up front suggests to me that this array would be longer.

Are you able to introduce a test that would fail without these changes?

---
not a request for this PR but food for thought

Also, when I am changing something but I don't think the code makes it clear what the results are after the change, I sometimes put in an extra commit up front

commit 1: add test where `expect()` has the values equal to the wrong value. (they are green) - or other tests that ensure that nothing will be broken.
commit 2: tests with new behavior and tests that show the `expect()` now returns a different value

Reply:
I should have pushed at the first commit to show how the added tests pass or fail.

Checking out the first commit and running the spec gives this.
Note that, out of the two added tests, only the one for array stye fails, and the hash one does not.

```console
shingo:~/lib/tmp/active_hash (stomk/yaml-alias-fields=) $ git co 062c7b
Note: switching to '062c7b'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 062c7b4 Add spec
shingo:~/lib/tmp/active_hash ((062c7b4...)) $ bundle exec rspec spec/active_yaml/aliases_spec.rb
....F......

Failures:

  1) ActiveYaml::Aliases using yaml arrays keys starting with "/" excludes them from fields
     Failure/Error: expect(model.field_names).to match_array [:name, :flavor, :price]
     
       expected collection contained:  [:flavor, :name, :price]
       actual collection contained:    [:"/aliases", :chip_flavor, :chip_price, :flavor, :name, :price, :soda_flavor, :soda_price]
       the extra elements were:        [:"/aliases", :chip_flavor, :chip_price, :soda_flavor, :soda_price]
     # ./spec/active_yaml/aliases_spec.rb:44:in `block (4 levels) in <top (required)>'

Deprecation Warnings:

Requiring `rspec/autorun` when running RSpec via the `rspec` command is deprecated. Called from /Users/shingo/lib/tmp/active_hash/spec/spec_helper.rb:4:in `require'.


If you need more of the backtrace for any of these deprecations to
identify where to make the necessary changes, you can configure
`config.raise_errors_for_deprecations!`, and it will turn the
deprecation warnings into errors, giving you the full backtrace.

1 deprecation warning total

Finished in 0.05192 seconds (files took 0.79868 seconds to load)
11 examples, 1 failure

Failed examples:

rspec ./spec/active_yaml/aliases_spec.rb:42 # ActiveYaml::Aliases using yaml arrays keys starting with "/" excludes them from fields

shingo:~/lib/tmp/active_hash ((062c7b4...)) $ 
```

Reply:
I'll explain what had happened before my changes.

Inserting some `binding.pry` in the code:

<details>

```diff
shingo:~/lib/tmp/active_hash (master *=) $ git diff
diff --git a/Gemfile b/Gemfile
index 8a435c8..028e855 100644
--- a/Gemfile
+++ b/Gemfile
@@ -7,6 +7,7 @@ gem 'wwtd'
 gem 'rake'
 gem 'test-unit'
 gem 'json'
+gem 'pry-byebug'
 
 platforms :jruby do
   gem 'activerecord-jdbcsqlite3-adapter', '>= 1.3.6'
diff --git a/lib/active_yaml/aliases.rb b/lib/active_yaml/aliases.rb
index c268a5e..b7a0684 100644
--- a/lib/active_yaml/aliases.rb
+++ b/lib/active_yaml/aliases.rb
@@ -12,6 +12,7 @@ module ActiveYaml
       end
 
       def raw_data
+        binding.pry
         super.reject do |k, v|
           v.kind_of? Hash and k.match(/^\//i)
         end
@@ -20,6 +21,7 @@ module ActiveYaml
     end
 
     def initialize(attributes={})
+      binding.pry
       super unless attributes.keys.index { |k| k.to_s.match(/^\//i) }
     end
   end
diff --git a/spec/spec_helper.rb b/spec/spec_helper.rb
index 6d9e03e..8fb1248 100644
--- a/spec/spec_helper.rb
+++ b/spec/spec_helper.rb
@@ -1,5 +1,6 @@
 require "bundler/setup"
 require "pry"
+require "pry-byebug"
 require 'rspec'
 require 'rspec/autorun'
 require 'yaml'
shingo:~/lib/tmp/active_hash (master *=) $ 
```

</details>

In hash style, the alias entries are filtered out up front in `ActiveYaml::Aliases.raw_data`, and the rejection in `initialize` does not take effect:

```console
shingo:~/lib/tmp/active_hash (master *=) $ bundle exec rspec spec/active_yaml/aliases_spec.rb:60
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/control_d_handler.rb:5: warning: control_d_handler's arity of 2 parameters was deprecated (eval_string, pry_instance). Now it gets passed just 1 parameter (pry_instance)
Run options: include {:locations=>{"./spec/active_yaml/aliases_spec.rb"=>[60]}}

From: /Users/shingo/lib/tmp/active_hash/lib/active_yaml/aliases.rb:16 ActiveYaml::Aliases::ClassMethods#raw_data:

    14: def raw_data
    15:   binding.pry
 => 16:   super.reject do |k, v|
    17:     v.kind_of? Hash and k.match(/^\//i)
    18:   end
    19: end

[1] pry(KeyProduct)> super
=> {"/aliases"=>{"soda_flavor"=>"sweet", "soda_price"=>1.0, "chip_flavor"=>"salty", "chip_price"=>1.5},
 "coke"=>{"id"=>1, "name"=>"Coke", "flavor"=>"sweet", "price"=>1.0},
 "pepsi"=>{"id"=>2, "name"=>"Pepsi", "flavor"=>"sweet", "price"=>1.0},
 "pringles"=>{"id"=>3, "name"=>"Pringles", "flavor"=>"salty", "price"=>1.5, "slogan"=>nil},
 "eta"=>{"id"=>4, "name"=>"ETA", "flavor"=>"salty", "price"=>1.5}}
[2] pry(KeyProduct)> super.reject do |k, v|  v.kind_of? Hash and k.match(/^\//i); end
=> {"coke"=>{"id"=>1, "name"=>"Coke", "flavor"=>"sweet", "price"=>1.0},
 "pepsi"=>{"id"=>2, "name"=>"Pepsi", "flavor"=>"sweet", "price"=>1.0},
 "pringles"=>{"id"=>3, "name"=>"Pringles", "flavor"=>"salty", "price"=>1.5, "slogan"=>nil},
 "eta"=>{"id"=>4, "name"=>"ETA", "flavor"=>"salty", "price"=>1.5}}
[3] pry(KeyProduct)> continue
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:13: warning: _pry_ is deprecated, use pry_instance instead
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:15: warning: _pry_ is deprecated, use pry_instance instead

From: /Users/shingo/lib/tmp/active_hash/lib/active_yaml/aliases.rb:25 ActiveYaml::Aliases#initialize:

    23: def initialize(attributes={})
    24:   binding.pry
 => 25:   super unless attributes.keys.index { |k| k.to_s.match(/^\//i) }
    26: end

[1] pry(#<KeyProduct>)> attributes
=> {:id=>1, :name=>"Coke", :flavor=>"sweet", :price=>1.0}
[2] pry(#<KeyProduct>)> attributes.keys.index { |k| k.to_s.match(/^\//i) }
=> nil
[3] pry(#<KeyProduct>)> continue
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:13: warning: _pry_ is deprecated, use pry_instance instead
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:15: warning: _pry_ is deprecated, use pry_instance instead

From: /Users/shingo/lib/tmp/active_hash/lib/active_yaml/aliases.rb:25 ActiveYaml::Aliases#initialize:

    23: def initialize(attributes={})
    24:   binding.pry
 => 25:   super unless attributes.keys.index { |k| k.to_s.match(/^\//i) }
    26: end

[1] pry(#<KeyProduct>)> attributes
=> {:id=>2, :name=>"Pepsi", :flavor=>"sweet", :price=>1.0}
[2] pry(#<KeyProduct>)> continue
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:13: warning: _pry_ is deprecated, use pry_instance instead
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:15: warning: _pry_ is deprecated, use pry_instance instead

From: /Users/shingo/lib/tmp/active_hash/lib/active_yaml/aliases.rb:25 ActiveYaml::Aliases#initialize:

    23: def initialize(attributes={})
    24:   binding.pry
 => 25:   super unless attributes.keys.index { |k| k.to_s.match(/^\//i) }
    26: end

[1] pry(#<KeyProduct>)> attributes
=> {:id=>3, :name=>"Pringles", :flavor=>"salty", :price=>1.5, :slogan=>nil}
[2] pry(#<KeyProduct>)> continue
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:13: warning: _pry_ is deprecated, use pry_instance instead
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:15: warning: _pry_ is deprecated, use pry_instance instead

From: /Users/shingo/lib/tmp/active_hash/lib/active_yaml/aliases.rb:25 ActiveYaml::Aliases#initialize:

    23: def initialize(attributes={})
    24:   binding.pry
 => 25:   super unless attributes.keys.index { |k| k.to_s.match(/^\//i) }
    26: end

[1] pry(#<KeyProduct>)> attributes
=> {:id=>4, :name=>"ETA", :flavor=>"salty", :price=>1.5}
[2] pry(#<KeyProduct>)> continue
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:13: warning: _pry_ is deprecated, use pry_instance instead
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:15: warning: _pry_ is deprecated, use pry_instance instead
.

Deprecation Warnings:

Requiring `rspec/autorun` when running RSpec via the `rspec` command is deprecated. Called from /Users/shingo/lib/tmp/active_hash/spec/spec_helper.rb:5:in `require'.


If you need more of the backtrace for any of these deprecations to
identify where to make the necessary changes, you can configure
`config.raise_errors_for_deprecations!`, and it will turn the
deprecation warnings into errors, giving you the full backtrace.

1 deprecation warning total

Finished in 41.46 seconds (files took 0.59325 seconds to load)
1 example, 0 failures

shingo:~/lib/tmp/active_hash (master *=) $ 
```

In array style, on the other hand, the alias entries are not filtered out in `raw_data`, but are rejected in `initialize` (called from `ActiveHash::Base.data=` just before `insert`).

https://github.com/zilkey/active_hash/blob/1a07f3ea7b4920f35843010cc515be8c8695e612/lib/active_hash/base.rb#L126

```console
shingo:~/lib/tmp/active_hash (master *=) $ bundle exec rspec spec/active_yaml/aliases_spec.rb:38
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/control_d_handler.rb:5: warning: control_d_handler's arity of 2 parameters was deprecated (eval_string, pry_instance). Now it gets passed just 1 parameter (pry_instance)
Run options: include {:locations=>{"./spec/active_yaml/aliases_spec.rb"=>[38]}}

From: /Users/shingo/lib/tmp/active_hash/lib/active_yaml/aliases.rb:16 ActiveYaml::Aliases::ClassMethods#raw_data:

    14: def raw_data
    15:   binding.pry
 => 16:   super.reject do |k, v|
    17:     v.kind_of? Hash and k.match(/^\//i)
    18:   end
    19: end

[1] pry(ArrayProduct)> super
=> [{"/aliases"=>nil, "soda_flavor"=>"sweet", "soda_price"=>1.0, "chip_flavor"=>"salty", "chip_price"=>1.5},
 {"id"=>1, "name"=>"Coke", "flavor"=>"sweet", "price"=>1.0},
 {"id"=>2, "name"=>"Pepsi", "flavor"=>"sweet", "price"=>1.0},
 {"id"=>3, "name"=>"Pringles", "flavor"=>"salty", "price"=>1.5},
 {"id"=>4, "name"=>"ETA", "flavor"=>"salty", "price"=>1.5}]
[2] pry(ArrayProduct)> super.reject do |k, v| v.kind_of? Hash and k.match(/^\//i); end
=> [{"/aliases"=>nil, "soda_flavor"=>"sweet", "soda_price"=>1.0, "chip_flavor"=>"salty", "chip_price"=>1.5},
 {"id"=>1, "name"=>"Coke", "flavor"=>"sweet", "price"=>1.0},
 {"id"=>2, "name"=>"Pepsi", "flavor"=>"sweet", "price"=>1.0},
 {"id"=>3, "name"=>"Pringles", "flavor"=>"salty", "price"=>1.5},
 {"id"=>4, "name"=>"ETA", "flavor"=>"salty", "price"=>1.5}]
[3] pry(ArrayProduct)> continue
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:13: warning: _pry_ is deprecated, use pry_instance instead
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:15: warning: _pry_ is deprecated, use pry_instance instead

From: /Users/shingo/lib/tmp/active_hash/lib/active_yaml/aliases.rb:25 ActiveYaml::Aliases#initialize:

    23: def initialize(attributes={})
    24:   binding.pry
 => 25:   super unless attributes.keys.index { |k| k.to_s.match(/^\//i) }
    26: end

[1] pry(#<ArrayProduct>)> attributes
=> {:"/aliases"=>nil, :soda_flavor=>"sweet", :soda_price=>1.0, :chip_flavor=>"salty", :chip_price=>1.5}
[2] pry(#<ArrayProduct>)> attributes.keys.index { |k| k.to_s.match(/^\//i) }
=> 0
[3] pry(#<ArrayProduct>)> !0
=> false
[4] pry(#<ArrayProduct>)> !1
=> false
[5] pry(#<ArrayProduct>)> continue
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:13: warning: _pry_ is deprecated, use pry_instance instead
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:15: warning: _pry_ is deprecated, use pry_instance instead

From: /Users/shingo/lib/tmp/active_hash/lib/active_yaml/aliases.rb:25 ActiveYaml::Aliases#initialize:

    23: def initialize(attributes={})
    24:   binding.pry
 => 25:   super unless attributes.keys.index { |k| k.to_s.match(/^\//i) }
    26: end

[1] pry(#<ArrayProduct>)> attributes
=> {:id=>1, :name=>"Coke", :flavor=>"sweet", :price=>1.0}
[2] pry(#<ArrayProduct>)> attributes.keys.index { |k| k.to_s.match(/^\//i) }
=> nil
[3] pry(#<ArrayProduct>)> continue
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:13: warning: _pry_ is deprecated, use pry_instance instead
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:15: warning: _pry_ is deprecated, use pry_instance instead

From: /Users/shingo/lib/tmp/active_hash/lib/active_yaml/aliases.rb:25 ActiveYaml::Aliases#initialize:

    23: def initialize(attributes={})
    24:   binding.pry
 => 25:   super unless attributes.keys.index { |k| k.to_s.match(/^\//i) }
    26: end

[1] pry(#<ArrayProduct>)> attributes
=> {:id=>2, :name=>"Pepsi", :flavor=>"sweet", :price=>1.0}
[2] pry(#<ArrayProduct>)> continue
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:13: warning: _pry_ is deprecated, use pry_instance instead
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:15: warning: _pry_ is deprecated, use pry_instance instead

From: /Users/shingo/lib/tmp/active_hash/lib/active_yaml/aliases.rb:25 ActiveYaml::Aliases#initialize:

    23: def initialize(attributes={})
    24:   binding.pry
 => 25:   super unless attributes.keys.index { |k| k.to_s.match(/^\//i) }
    26: end

[1] pry(#<ArrayProduct>)> attributes
=> {:id=>3, :name=>"Pringles", :flavor=>"salty", :price=>1.5}
[2] pry(#<ArrayProduct>)> continue
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:13: warning: _pry_ is deprecated, use pry_instance instead
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:15: warning: _pry_ is deprecated, use pry_instance instead

From: /Users/shingo/lib/tmp/active_hash/lib/active_yaml/aliases.rb:25 ActiveYaml::Aliases#initialize:

    23: def initialize(attributes={})
    24:   binding.pry
 => 25:   super unless attributes.keys.index { |k| k.to_s.match(/^\//i) }
    26: end

[1] pry(#<ArrayProduct>)> attributes
=> {:id=>4, :name=>"ETA", :flavor=>"salty", :price=>1.5}
[2] pry(#<ArrayProduct>)> continue
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:13: warning: _pry_ is deprecated, use pry_instance instead
/Users/shingo/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/pry-byebug-3.7.0/lib/pry-byebug/helpers/navigation.rb:15: warning: _pry_ is deprecated, use pry_instance instead
.

Deprecation Warnings:

Requiring `rspec/autorun` when running RSpec via the `rspec` command is deprecated. Called from /Users/shingo/lib/tmp/active_hash/spec/spec_helper.rb:5:in `require'.


If you need more of the backtrace for any of these deprecations to
identify where to make the necessary changes, you can configure
`config.raise_errors_for_deprecations!`, and it will turn the
deprecation warnings into errors, giving you the full backtrace.

1 deprecation warning total

Finished in 2 minutes 44.3 seconds (files took 1.02 seconds to load)
1 example, 0 failures

shingo:~/lib/tmp/active_hash (master *=) $ 
```

So what happens here is, in array style, the alias entries are passed to `data=` and their fields are registered by `auto_assign_fields`, while the record itself is rejected at `insert new(hash)`.

In hash style, since the alias entries are rejected up front and not passed to `data=`, both fields and record are not registered.

https://github.com/zilkey/active_hash/blob/1a07f3ea7b4920f35843010cc515be8c8695e612/lib/active_hash/base.rb#L118-L129

Reply:
Explaning above, I noticed that the `initialize` rejection is no longer necessary because alias entries are now always rejected up front in `raw_data`. I'll remove it.

Reply:
No, the junk data won't get into `_data` because it is already rejected and is not passed to `data=`.

Reply:
thanks for your patience - your previous quest cleared this up as well